libs/@local/harpc/server/src/router.rs

```diff
@@ -0,0 +1,282 @@
+use alloc::sync::Arc;
+use core::{
+    future::{self, Ready},
+    task::{Context, Poll},
+};
+
+use frunk::{HCons, HNil};
+use futures::{
+    FutureExt,
+    future::{Either, Map},
+};
+use harpc_net::codec::ErrorEncoder;
+use harpc_service::delegate::ServiceDelegate;
+use harpc_tower::{
+    body::{Body, BodyExt},
+    request::Request,
+    response::{BoxedResponse, Parts, Response},
+};
+use harpc_types::{service::ServiceId, version::Version};
+use tower::{Layer, Service, ServiceBuilder, ServiceExt, layer::util::Identity, util::Oneshot};
+
+use crate::{delegate::ServiceDelegateHandler, error::NotFound, session::SessionStorage};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Handler<S> {
+    service: ServiceId,
+    version: Version,
+
+    inner: S,
+}
+
+/// Route requests to the appropriate handler based on the request's service and version.
+///
+/// This is essentially a type-level linked list of handlers, it's boxed equivalent is [`Steer`],
+/// but unlike [`Steer`] it doesn't require the same type for all handlers and a separation of both
+/// the meta information (service id and version) and the actual handler. Unlike [`Steer`], it also
+/// doesn't require `&mut self` access, which allows for more granular cloning.
+///
+/// # Design motivations
+///
+/// The reason why we essentially require `Clone`/`Copy` for the handlers is that once the route is
+/// constructed it needs to be available for each request that happens, now, there are multiple ways
+/// to achieve this.
+///
+/// One way would be to simply clone the entire [`Router`] during each request, but that has the
+/// downside of potentially cloning a lot of data that isn't actually required for each request,
+/// making the addition of new handlers slower and slower.
+/// The other solution instead would be to `Mutex` the entire [`Router`], but that would make the
+/// entire [`Router`] essentially single-threaded, which is not ideal.
+///
+/// This takes a middle ground, which is similar in implementation to other tower-based frameworks,
+/// such as axum. The inner routes are stored in an `Arc<T>`, which is cheap to clone, but means we
+/// need to require `&self` during routing. Once a route was chosen, we simply clone the service
+/// (and oneshot) the service. This keeps the cloned data minimal and allows for multi-threading.
+///
+/// The downside is that we're unable to keep any state in a service delegate that's persisted
+/// across invocations. To signal this, `ServiceDelegate` takes `self` and consumes it (even though
+/// it isn't strictly needed), as well as the use of sessions. To store any information across
+/// calls, one must make use of smart pointers, such as `Arc`.
+///
+/// [`Steer`]: tower::steer::Steer
```

Comment:
Well
<img width="261" alt="image" src="https://github.com/user-attachments/assets/0cc1e3e5-c82d-4bca-856b-84b577b51ec3">
```suggestion
/// [`Steer`]: https://docs.rs/tower/latest/tower/steer/struct.Steer.html
```